illumos-utils/src/running_zone.rs

```diff
@@ -257,42 +267,95 @@ impl RunningZone {
         let addrobj =
             AddrObject::new(port.vnic_name(), name).map_err(|err| {
                 EnsureAddressError::AddrObject {
-                    request: addrtype,
+                    request: AddressRequest::Dhcp,
                     zone: self.inner.name.clone(),
                     err,
                 }
             })?;
-        let network =
-            Zones::ensure_address(Some(&self.inner.name), &addrobj, addrtype)?;
-        if let AddressRequest::Dhcp = addrtype {
+        let zone = Some(self.inner.name.as_ref());
+        if let IpAddr::V4(gateway) = port.gateway().ip() {
+            let addr =
+                Zones::ensure_address(zone, &addrobj, AddressRequest::Dhcp)?;
             // TODO-remove: OPTE's DHCP "server" returns the list of routes to add
             // via option 121 (Classless Static Route). The illumos DHCP client
             // currently does not support this option, so we add the routes
             // manually here.
             // https://www.illumos.org/issues/11990
-            if let IpAddr::V4(gateway) = port.gateway().ip() {
-                let gateway_ip = gateway.to_string();
-                let private_ip = network.ip();
-                self.run_cmd(&[
-                    "/usr/sbin/route",
-                    "add",
-                    "-host",
-                    &gateway_ip,
-                    &private_ip.to_string(),
-                    "-interface",
-                    "-ifp",
-                    port.vnic_name(),
-                ])?;
-                self.run_cmd(&[
-                    "/usr/sbin/route",
-                    "add",
-                    "-inet",
-                    "default",
-                    &gateway_ip,
-                ])?;
-            };
+            let gateway_ip = gateway.to_string();
+            let private_ip = addr.ip();
+            self.run_cmd(&[
+                "/usr/sbin/route",
+                "add",
+                "-host",
+                &gateway_ip,
+                &private_ip.to_string(),
+                "-interface",
+                "-ifp",
+                port.vnic_name(),
+            ])?;
+            self.run_cmd(&[
+                "/usr/sbin/route",
+                "add",
+                "-inet",
+                "default",
+                &gateway_ip,
+            ])?;
+            Ok(addr)
+        } else {
+            // If the port is using IPv6 addressing we still want it to use
+            // DHCP(v6) which requires first creating a link-local address.
+            Zones::ensure_has_link_local_v6_address(zone, &addrobj).map_err(
```

Comment:
I think this need another addrobj with a different name, as my zones see duplicates:
```
root@oxz_nexus:~# ipadm
ADDROBJ           TYPE     STATE        ADDR
lo0/v4            static   ok           127.0.0.1/8
lo0/v6            static   ok           ::1/128
oxControlService4/ll addrconf ok        fe80::8:20ff:fe71:c5d2/10
oxControlService4/omicron6 static ok    fd00:1122:3344:102::4/64
vopte1/public     addrconf ok           fe80::aa40:25ff:feff:fd45/10
vopte1/public     addrconf ok           fd77:e9d2:9cd9:1::5/128
```

Reply:
So, the second gets added automatically and I don't know if it's possible to have them be different? Basically the flow is:
1. Link local addrconf is created (`fe80::aa40:25ff:feff:fd45/10`)
2. Router Solicitation is sent out with source being the link-local address and destination being the All-Routers IPv6 Multicast Address (`ff02::2`)
3. OPTE will generate a Router Advertisement reply with the "managed" bit set indicating the requester needs to use DHCPv6 to acquire an address.
4. `dhcpagent` then goes through the DHCPv6 solicit/advertise/request/reply process to acquire the non link-local address originally configured in the VPC (`fd77:e9d2:9cd9:1::5/128`).

I'm assuming 4 is where that second addrobj gets added and so I don't think we can change the name (nor I guess should we?)